### ISSUE TITLE
Refreshing auth tokens for SignalR

### ISSUE THREAD

**Issue by analogrelay** [dotnet/aspnetcore#5297]
Below is one option we've considered, but I'm re-framing this issue to build *some* story for refreshing auth tokens.

To improve the ability to "refresh" expired tokens, we should consider caching the access token provided by the factory. Then, when an HTTP request gets a `401`, we call the factory again before re-issuing the request. That way the user can configure a process to "refresh" the token without forcing the connection to be reestablished

* For the WebSockets transport, this has no effect. There is only ever a single request. This logic would **not** cover reconnecting in the event of something like #1159 (where the WebSocket is terminated when the token expires)
* For the SSE transport, this only affects `POST` (send) requests. We would call the token factory again and re-issue the send. The unsent data would stay buffered in the pipe
* For Long Polling, this affects `POST` requests like SSE, and the `GET` (poll) requests. The client would assume that a `401` error indicates that the data is **still in the pipe** for them to read. The server would be expected to keep data in the pipe in the case of a `401`

**Comment by anurse** [comment]
@SteveSandersonMS @rynowak FYI, just something that popped up in our backlog grooming and @BrennanConroy mentioned this was something you cared about. So we put our special happy label on it.

**Comment by anurse** [comment]
Expanding this to also cover the possibility of in-band refresh of the token. We may want to build a way to refresh the user principal **without** terminating the connection.

**Comment by BrennanConroy** [comment]
We should author a doc about handling auth expiration correctly. See https://github.com/dotnet/aspnetcore/issues/5283#issuecomment-368864944 for a code example.

@bradygaster asked me to add @IEvangelist

**Comment by BrennanConroy** [comment]
Add an option to enable automatic refresh of connection on token expiration (note: this closes the connection)
* Store some sort of auth information that has expiration on it on the transport
* Use IConnectionLifetimeNotificationFeature from heartbeat on expiration to tell SignalR to abort connection (with retry allowed)
* If expiration of token happens, can we first run auth again to check if the token hasn't had its expiry updated?

**Comment by SebastianKunz** [comment]
Not checking the expiration of the access token in SignalR is a huge security risk. And it adds some work for the appilcation developer to make sure the access token is always fresh. I think it makes sense to add it to the SignalR library where the developer can opt in the feature #14578
I don't like the approach of closing the connection to then refresh the access token and then connect with the new access token. It does the job, but it's slow and hacky. There is also the possibility that the backend gets confused because of a completely new connectionId.

Furthermore I see a second problem arising, which is not solved by keeping the access token fresh. And that is changing roles within the access token.
For that let me give an example.

You have a priviliged user that manages other users in your application. The priviliged user then removes a role from one of the users he manages (e.g. removes the role of adding comment to you blogpost). Assuming this communication is via SignalR and the access token is always valid (not expired) then the user is still authorized to the the tasks (comment on your blogpost) he just was removed from. In the worst case, the access token was just refreshed and the user can abuse his roles up to EXPIRATION_TIME (e.g. 4 minutes).

This is a more complex problem to solve then just making sure that the token is always valid, because the Identity Provider needs to inform you that the roles of a user were updated.

To sum this up. 
I see two problems with signalR and authorization. The first is that the access token is never looked at after the inital connection and second updating roles while a connection is live.
I think we should focus on the first one (keeping the access token fresh). I think it would already help a lot of developers when SignalR with websockets support this feature.

**Comment by JoFrMueller** [comment]
@BrennanConroy killing the connection in case of an expired refresh token sounds like a blunt hack to be honest. Everything that relates to the connection would be thrown away, while what we actually want to achieve is a "cache invalidation" on the side of the hub with the support of the client. Couldn't we think about it from the "bi-directional communication feature" perspective of SignalR itself?

1. Hub informs clients some seconds (configurable) before the access token will become invalid. Either because of a timer or because of the heartbeat mechanism. The hub has to know about it, because it persists/caches the initial access token and relates it to the connection itself.
2. Clients can react to this event, that is triggered by the hub to then re-send their access token (which should be refreshed by their local access token handling mechanism anyways) to allow the hub to update the authorization properties related to the client.
3. The hub receives the updated token from the client and overwrites all related ClaimsPrincipal properties for the sending client and therefore updates its internally cached/connectation related state to reflect the latest acces token content.
4. Because of this update, if we remove a role from a client (like described by @SebastianKunz f.e.), this change will be reflected with the next access token refresh cycle, which usually should take 5 minutes maximum.

**Comment by BrennanConroy** [comment]
Apologies, I realize I left out some critical information when jotting down notes. Closing the connection after token expiration is one part of what we plan on implementing. We do want to have a nicer experience where the token can be refreshed without closing the connection. That does require a lot more thought and care and will be additive with the expiration check.

The current plan is to add just the expiration part in 6.0 (opt in) and then build on top of that and design the experience for refresh for a future release.

**Comment by BrennanConroy** [comment]
6.0 work is done, there is a new option to close connections on auth expiration. See https://github.com/dotnet/aspnetcore/pull/32431 for info.

Backlogging for future work in 7.0

**Comment by gregoriusus** [comment]
In addition to the security flaw mentioned before, there is also another problem. In high loaded signalr system disconnecting and reconnecting can lead to a lot of messages being lost. Disconnecting and reconnecting are expensive operations that should be avoided.

**Comment by gregoriusus** [comment]
I am testing CloseOnAuthenticationExpiration feature. It works. On my development machine with all resources free it takes 1 second to process all operations:
[2022-02-02T23:19:24.859Z] Information: Connection disconnected.
[2022-02-02T23:19:25.904Z] Information: WebSocket connected to....

A lot can happen in 1 second and on a high load system with a lot of users and messages can lead to messages lost.

**Comment by rafikiassumani-msft** [comment]
Below are individual items to consider for refreshing auth tokens for SignalR: 

- [ ] Fix error message on client side for OnClose (.NET7):

* Add info to CloseMessage that auth is expired and make it easy for clients to see that (strongly typed exception) and request a new token if required. This should include the error message and status code (401 for example). 

- [ ] Protocol change (.NET8):

* Add new hub message type for "pre-close" that's effectively an auth challenge and has a grace period for a response before killing connection
* Could make use of Hub message headers to flow auth info to server instead.

- [ ] Immediate easy ™ win for (.NET7):

* Change AccessTokenFactory to only be called on startup and on 401/403 and retry the request, not for every single http request

- [ ] Doc improvement (.NET7):

* Add doc on how users can setup an endpoint to refresh their token on the server

**Comment by davidfowl** [comment]
This is a popular issue, so I figured I'd shared some updates. The team is looking at ways to keep the make sure there's a new auth token that represents the updated user information. Since SignalR isn't a simple request-response based system, it's challenging and anything we do here will require protocol changes (that means new client and server changes). 

There are 2 approaches being considered:

1. A way to renew user information using the [The transport protocol](https://github.com/dotnet/aspnetcore/blob/f543e3552514c5c420eeddd55c505bbc131f10a6/src/SignalR/docs/specs/TransportProtocols.md) (HTTP in the common case) 
2. A way to renew the user information in the [The hub protocol](https://github.com/dotnet/aspnetcore/blob/f543e3552514c5c420eeddd55c505bbc131f10a6/src/SignalR/docs/specs/HubProtocol.md)

**Comment by Gruski** [comment]
I'd be happy if we could control when the accessTokenFactory is called to retrieve new access token. For custom authentication I would like to send a new auth token on every hub method call for reasons too elaborate to go into detail here. Sure I could do it though method parameters but that's ugly especially when there is an elegant way to do it on initial connection. Then on the back end we could develop more unified ways to custom handle those tokens, update identity, and so on.

**Comment by gregoriusus** [comment]
Has been this addressed in. NET8?

**Comment by DeepWorksStudios** [comment]
> Has been this addressed in. NET8? At the moment, when getting a new token with refresh token, I can only disconnect an connect with a new token to trigger all auth procedure in Signalr backend. Isn't this a big issue? Can't just implement client side function who will send some magic word with token and reauthenticate connection or change URL param?

I agree this would be required to implement in net8 i'm having the same issues

**Comment by gregoriusus** [comment]
There are at least 3 issues ( opened on this issue. Any movement on this directions or at least feedback?

**Comment by Misiu** [comment]
This issue should be moved to .NET 9 Planning as .NET 8 will be released soon. Hopefully, it will get addressed in .NET 9

**Comment by Skyppid** [comment]
> This issue should be moved to .NET 9 Planning as .NET 8 will be released soon. Hopefully, it will get addressed in .NET 9

As often as it was moved to the next iteration without **any** visible update on it makes me doubt that .NET 9 will be our saviour. It's been two years almost since I last used it and had to work around this issue. Now again, I really thought they might have fixed it by now. But looks like even ASPNET.Core and SignalR aren't important enough projects for MS to fix critical issues. Sad.

**Comment by gregoriusus** [comment]
As I am aware, there is even no clue how to solve this from architecture point of view

**Comment by gregoriusus** [comment]
Is this issue's milestone moved to NET8 planning?

**Comment by DeepWorksStudios** [comment]
It's just to sad that signalR suffer from that issue and since months I hoped that they fix it but I guess it will never be fixed my impression is that all those Microsoft's projects are suffering from the same issues that the team behind it has not the resources to tackle all the bleeding wounds in their project and they just try to fix it but it's simply to much work for those people. That's my opinion based on the experience working at multiple .net Maui and asp.net core projects. Reported over 12 issues and followed a lot more of already reported issues that I as well encountered in those projects. And still a lot of the issues are still remaining present in the latest versions.

**Comment by gregoriusus** [comment]
@mkArtakMSFT, guess you have some insider information. Are you going to tackle this issue?

**Comment by davidfowl** [comment]
Hopefully in .NET 9

**Comment by gregoriusus** [comment]
@davidfowl , are there any new information regarding 2 approaches, have you talked about it?

**Comment by Skyppid** [comment]
> Hopefully in .NET 9

That'd be great. SignalR itself is a great tool and usually works really well. But that issue is a huge deal breaker and always was. Is about time that this will be fixed.

Maybe the community can help with that.

**Comment by yuvi-pc** [comment]
Came across this [stackoverflow ](https://stackoverflow.com/a/70871191/4060087) workaround. It is working for us for our self hosted service. Is this a viable solution for now?

**Comment by gregoriusus** [comment]
> Came across this [stackoverflow ](https://stackoverflow.com/a/70871191/4060087) workaround. It is working for us for our self hosted service. Is this a viable solution for now?

Well, this implies to stop and reconnect... this is only solution for now....

**Comment by davidfowl** [comment]
> @davidfowl , are there any new information regarding 2 approaches, have you talked about it?

Yes, we would go with an "over the transport" approach. Which means protocol changes to accommodate doing auth without a reconnect.

**Comment by gregoriusus** [comment]
That's great. Is it already on road map?

**Comment by gregoriusus** [comment]
Any news on this topic?

**Comment by davidfowl** [comment]
> So maybe .NET 10?

At a minimum

**Comment by Skyppid** [comment]
Wow that's crazy... too bad there's not much good alternatives to SignalR as of now...

**Comment by reduckted** [comment]
> At a minimum

The earliest we'll see it is in the next version? How about a maximum? 😆

**Comment by DeepWorksStudios** [comment]
> > At a minimum
> 
> 
> 
> The earliest we'll see it is in the next version? How about a maximum? 😆 

It's sadly the fact that Microsoft does not invest enough in the development department in comparison to how many technologies they are invented and maintaining. 

Best course is to build the solution yourself or even better try to avoid such incompleted frameworks.

**Comment by thirstyape** [comment]
6 years and 9 months and not even close to any real progress, just a bunch of classic MS bureaucracy.

Why is there not at least a manual way to refresh tokens or even just the stored value?

**Comment by Skyppid** [comment]
> 6 years and 9 months and not even close to any real progress, just a bunch of classic MS bureaucracy.
> 
> Why is there not at least a manual way to refresh tokens or even just the stored value?

From what I understand by using the `accessTokenFactory` you always fetch a fresh token when a request is being sended. To optimize it, you can cache it until expiry and request a new one only if expired.

```ts
    const hubConnectionBuilder = new HubConnectionBuilder().withUrl(this.SERVER_HUB_URL, {
      transport: HttpTransportType.None,
      logger: LogLevel.Debug,
      accessTokenFactory: () => this.identityService.getToken()
    });
```

**Comment by gregoriusus** [comment]
Interesting that this is not some major issue. No one is refreshing token until logout/login?

**Comment by Skyppid** [comment]
Is it? I remember in a previous company we used it and had the problem that after the expiry all requests failed. Back then we had to actually reconnect. But using the factory it requests a token on each request, so if the token is outdated it should still fail?

Or do you mean backend-side? Honestly, we haven't been using SignalR that much so far, as those features were just UX convenience and there were more pressing things to build. It works for us, but maybe I'm overlooking something here as I'm not that deep into it anymore.

Just wanted to hint that this helps at least with some common issues in low-frequency / non-critical scenarios.

**Comment by thirstyape** [comment]
> Interesting that this is not some major issue. No one is refreshing token until logout/login?

By default this is pretty much the issue, the token request does not happen until an HTTP request goes out, which doesn't happen often as most communication is facilitated via WebSocket. 

Maybe the thought process here is that since the connection remains open it doesn't require re-validation? Not sure I agree with this.

Just seems wrong when I'm adding methods on the hub for them to not check for a valid token on each request. And when an invalid token is found just abort the connection. My current workaround is two-fold: on the hub add a method that informs whether the token is valid; on the client check with the hub each request whether the token is valid and reconnect when it is not. Additionally, the hub will do a single retry if the request failed due to closure from the hub.

**Comment by thirstyape** [comment]
> Yes, we are solving now, on client side, to pool all request to signalr when we are refreshing token and then we disconnect and reconnect. But on high volume of calls to/from signalr this will pose a problem of messages being lost or invest on some kind of retry mechanism.

Exactly!

Only dealing with maybe 100 low traffic connections at a time right now, but this is going to be a scaling nightmare.

**Comment by DeepWorksStudios** [comment]
> Yes, we are solving now, on client side, to pool all request to signalr when we are refreshing token and then we disconnect and reconnect. But on high volume of calls to/from signalr this will pose a problem of messages being lost or invest on some kind of retry mechanism. 

It's more about the actual backend issue that we can have the case an invalid token or a user that changed permissions on runtime can still access protected methods since the validation just happens on the initial handshake of the websocket connection afterwards if the connection persists than we got the issue with unauthorized users can still access protected resources. The only workaround is to attach to the signalR heartbeat and validate it yourself but on client side we got the reconnect issue it's a slight different issue but still all is kinda sharing the same issue that authentication is rather fragile in signalR 

The workaround works but ain't optimal and causes scaling issues if the demand increases so yeah total conclusion is rather bad since it's a issue known for nearly 7 years

**Comment by DeepWorksStudios** [comment]
> > Yes, we are solving now, on client side, to pool all request to signalr when we are refreshing token and then we disconnect and reconnect. But on high volume of calls to/from signalr this will pose a problem of messages being lost or invest on some kind of retry mechanism.
> 
> Exactly!
> 
> Only dealing with maybe 100 low traffic connections at a time right now, but this is going to be a scaling nightmare. 

Sadly we already experienced issues  at our company in therms of scaling and there is no real solution sadly since it's a concept issue in signalR and can't be changed that easily that's why I wonder what did Microsoft did all the 6 years. That would be sufficient time to fix such a major flaw

**Comment by gregoriusus** [comment]
Me niether, this is additional issue :-(

At the time when they architect this, refresh token was not a big issue. We have one solution on old Signalr (NET framework) and we are using cookie based authentication. It works flawless. Anyone knows, if it is possible to mix and for signalr send cookie based authneticaton with some custom validation on server and for API calls token based?

**Comment by DeepWorksStudios** [comment]
> > It's more about the actual backend issue that we can have the case an invalid token or a user that changed permissions on runtime can still access protected methods since the validation just happens on the initial handshake of the websocket connection afterwards if the connection persists than we got the issue with unauthorized users can still access protected resources. The only workaround is to attach to the signalR heartbeat and validate it yourself but on client side we got the reconnect issue it's a slight different issue but still all is kinda sharing the same issue that authentication is rather fragile in signalR
> > 
> > The workaround works but ain't optimal and causes scaling issues if the demand increases so yeah total conclusion is rather bad since it's a issue known for nearly 7 years
> 
> Wow! I hadn't even thought of token invalidation or permission changes, that's huge.

It's a huge issue and no real solution wich would solve the root issue 

I mean that's why nobody srs uses Microsoft tech because their them self don't use it in the extend their should to prove it's viable for production and sadly this case if you require high security that's a no no 

It's a killer issue out of my perspective I wonder why nobody points that out as what it is 

I mean I follow this issue since 4 years and still no progress at all only more people find out about the surface issue but only a small part ion understands the overall impact on therms of auth security

**Comment by gregoriusus** [comment]
Do you think that token authentication for api calls and cookie based for signalr would help in this issue? Just a thought...

**Comment by thirstyape** [comment]
> Do you think that token authentication for api calls and cookie based for signalr would help in this issue? Just a thought...

From what I can tell this would be subject to the same issue. 

The documentation, and examples I googled, seem to suggest that the cookie would only be provided on the initial HTTP request, and subsequent requests over the WebSocket would be doing the same thing as token based requests.

**Comment by DeepWorksStudios** [comment]
> It has been a while since I have looked at the SignalR code in my project, but I think I mitigate the many SignalR issues/concerns with the following approach:
> 
> 1. My  SignalR access token is a very short-lived token just for the hubs - no other endpoints will accept them - they are different from my normal api endpoint access tokens.
> 
> 2. I never send data to the hub via the SignalR connection!  When calling the backend, I always use a normal api endpoint with the normal session security mechanism.
> 
> 3. As you can guess, data is only sent from the backend to the frontend, for the specified user guid that was part of the original access token when initially connecting to SignalR.  The backend can check the current role of the user (e.g. against the database) before sending the data if necessary or if concerned about a role change.
> I believe that makes it okay for the SignalR connection to remain open as long as the user is on that web page.
> If the user's web session times out, they are redirected to the login page, so their connection is closed from the frontend at that point.
> 
> 4. Due to the javascript library having issues reconnecting server-sent events or long-polling (more specifically not bothering to properly inspect the response status code), I use a custom HttpClient that extends DefaultHttpClient and overrides send()  in order to catch any error and and return a specific HttpResponse to address those issues to allow them to reconnect.
> 
> Having said all that, if I recall correctly (please be kind - I could be wrong!), the backend Hub Context keeps the initial connection request details - so the user's guid that was in the SignalR access token is always accessible via a claim via the Hub.Context.GetHttpContext().  I believe, if I wanted to, I could add other custom claims such as a session expiry timestamp if I wanted to check that later in a hub method, and then call Abort() if it had expired - or something similar if there was a role change and it had to be invalidated. Could be wrong!  
> Either way, given my approach, I don't need that.
> 
> Hope this helps someone.

Sadly it won't work for my use case but I see how it can minimize the issues for many use cases but still they are just a workaround out of my perspective

**Comment by Gruski** [comment]
I need this a few years ago too but since there was nothing being done I ended up changing my auth from token based to where info was actually stored in token to database based so now token never needs to be updated because I can just update the auth of the static client token on the server side in DB user's record. That also helped with heterogeneous systems that could not share auth which I also needed.

**Comment by DeepWorksStudios** [comment]
https://www.youtube.com/watch?v=kAGlATRcgY4
Security & Auth .NET 10 planning discussion

Thats currently live on going maybe it will tackel this issue in .net 10
At least i hope so

**Comment by DeepWorksStudios** [comment]
They mentioned that it will be worked on this issue in .Net 10 

Gladly 🎉

**Comment by Vandersteen** [comment]
Looks like they removed it from the list

<img width="385" alt="Image" src="https://github.com/user-attachments/assets/64845a56-76de-48f0-b5e3-dccf1b90346f" />

**Comment by Misiu** [comment]
It is still in .NET 10 Planning, so I hope this will be finally added. Fingers crossed.

**Comment by s-beji** [comment]
Fml I was really hoping they tackle that this time it's just disappointing

**Comment by Skyppid** [comment]
> It is still in .NET 10 Planning, so I hope this will be finally added. Fingers crossed.

.NET is already in RC - this won't be coming. At least the 2 weeks of workaround were now well invested and not just something we're gonna rip out in 3 months. Still it's hilarious how neglected this issue is.

SignalR has been spamming our Sentry for weeks now with thousands of incidents as once the token expires any disconnect leads to an exception due to user info not being valid anymore.

Now we signal to the backend services that a user reconnects for a new token, then we cut the connection but keep the state and reconnect. Acknowledgement happens, so in case it connects to a different pod the original knows it was handled. It's like 4 times the size of what our current hub actually handles feature-wise just to make the connection work.

I'd like to know why they keep discarding this issue all the time. It's a severe flaw. Would it require to rebuild SignalR from scratch or do they lack enough experienced developers to work out a solution for that? It's not quantum computing, is it?

**Comment by martincostello** [comment]
> It is still in .NET 10 Planning, so I hope this will be finally added

Given many other issues against the .NET 10 milestone have already been moved back to Backlog for lack of capacity I imagine this same is the case for this issue and it's just that the milestone hasn't been updated.

> .NET is already in RC

This isn't the case - that's usually in August. The next release for 10 is preview.5, and I imagine we'll get preview.6 in July.

**Comment by Skyppid** [comment]
> > .NET is already in RC
> 
> This isn't the case - that's usually in August. The next release for 10 is preview.5, and I imagine we'll get preview.6 in July.

True, my bad, mixed up Preview and RC. But still, no doubt that this won't come in .NET 10 as well. Who knows when though, as often as they moved it out of the roadmap I doubt it's gonna be in any of the next versions though. Well it works now and we won't add any features on top of it until it's fixed. Rather switch to a different solution alltogether.

**Comment by gregoriusus** [comment]
But what is other solution? socket.io?

**Comment by Skyppid** [comment]
That, raw websockets (unbearable amount of work probably) or gRPC Streaming. Unfortunately not a lot of good alternatives. That's why it's even worse that MS is not fixing the most crucial and basic issues of SignalR since it's actually a great tool.

**Comment by gregoriusus** [comment]
Well, at least other solutions (socket.io, MQTTnet), have ability to change headers on server side. Headers are set at handshake and remain same for whole session. But then one can make custom method resfresh-token and refresh token header on server side.

**Comment by Misiu** [comment]
So maybe .NET11? There is .NET 11 Planning Milestone.

**Comment by Skyppid** [comment]
We will certainly get rid of it. Seems like even though it's a great tool in theory the overall usage of it is too low for MS to care about maintaining it or even fixing the most crucial of design flaws. It's been known for years, makes it completely useless in almost any real use case and yet it get's ditched from the roadmap year by year.

Building on broken foundations.

**Comment by Gruski** [comment]
> We will certainly get rid of it. 

Get rid of what? SignalR? The thread is about adding a feature to it.

**Comment by Skyppid** [comment]
> > We will certainly get rid of it.
> 
> Get rid of what? SignalR? The thread is about adding a feature to it.

Sorry, I'm just frustrated. I mean this feature request exists for 7 years and was officially advertised to be addressed in .NET 10. Now it's back to Backlog, not even .NET 11. It's a fundamental flaw that makes SignalR useless in almost any production scenario or causes a painful amount of boilerplate to manually fix with plenty of issues coming with that.

I'll shut up now, I get that it's the wrong place to vent and am sorry for that. Still, it's really just frustrating. But we will ditch SignalR. There are other solutions and by now I lost any hope and regret the time we spent builing on top of it.

**Comment by MichaelHochriegl** [comment]
So glad to see this on the .Net11 planning board, I really hope it gets delivered this time, as Auth with Blazor Server is really cumbersome and I suppose this does not only target plain SignalR and will also be working in the Blazoer Server backed SignalR connection?

**Comment by davidfowl** [comment]
How are you thinking this helps with blazor server auth? What is painful today?

**Comment by MichaelHochriegl** [comment]
Blazor Server is really painful with Refresh Tokens. We have Keycloak as an IDP and get a short lived Access Token from it. We use the `RevalidatingServerAuthenticationStateProvider` to get a new Access Token with the Refresh Token, however, it is not really well integrated with the IDP configuration, so we have manual calls in there to the endpoints.

**Comment by reduckted** [comment]
> So glad to see this on the .Net11 planning board

I wouldn't get your hopes up. Based on the history, and given that .NET 11 Preview came out just the other day, I think it's safe to say this will be moved to .NET 12 planning.

**Comment by BrennanConroy** [comment]
> [!NOTE]
> ⚠️ Early prototype - design and API are subject to change, and this feature is not guaranteed to ship in .NET 11. We're sharing it to show direction and gather feedback. If there are scenarios below that aren't covered, or parts of the design you think aren't needed (or are missing), please let us know.

**Refreshing SignalR auth tokens without dropping the connection**

We've been prototyping auth token refresh for SignalR. The goal is to let a long-lived connection pick up a fresh token, and any authorization changes that come with it, without a disconnect/reconnect cycle.

**The problem**

Today a connection's token is only validated at the handshake. For long-lived connections that means:

* Token expiration on long-running connections (dashboards, chat, Blazor Server) forces a reconnect, which loses in-flight messages and adds load.
* Permission/role changes (e.g. an admin revokes a role) aren't picked up until the connection is torn down - it keeps running with stale claims.
* `CloseOnAuthenticationExpiration` detects expiry but still forces a full reconnect (~1s + message loss), which doesn't suit high-frequency scenarios.

**What we're planning**

* **Out-of-band /refresh endpoint** mapped next to /negotiate. The client sends a fresh token; the server re-authenticates and swaps the connection's ClaimsPrincipal in place - no transport restart.
* **Client-driven TTL countdown**. The server returns tokenLifetimeSeconds on the negotiate and refresh responses; the client counts down and refreshes early. No token parsing and no client/server clock-sync assumptions, so it works for JWT, opaque tokens, cookies, etc.
* **App-defined policy hook** (`OnAuthRefresh`). Apps can inspect the previous/new principal and approve or deny a refresh (e.g. enforce same user, downgrade-only permissions).
* **Hub notification** (`Hub.OnAuthRefreshedAsync`). Hub authors get a callback after a refresh to react/re-evaluate per-connection state, re-subscribe groups, etc.
* **Authorization re-evaluated automatically**. After a refresh, `[Authorize]` policies on hub methods run against the new claims, so revoked permissions take effect on the live connection.
* **User-identity changes are handled**. If a refresh changes the SignalR user id, the connection is re-keyed so `Clients.User(newId)` reaches it (and the old id no longer does), including across Redis/scaleout backplanes.
* **Consistent across transports**, including the Long Polling poll path, with a configurable grace period so an in-flight refresh can land before a connection is closed on expiration.

**Roughly what the API looks like**

**Server**

* `HttpConnectionDispatcherOptions.EnableAuthRefresh`, `AuthRefreshGracePeriod`, and an `OnAuthRefresh` callback (`AuthRefreshContext` with the previous/new principal, new expiration, and a deny reason).
* `Hub.OnAuthRefreshedAsync()` to react to a refresh.

**.NET client**

* `hubConnectionBuilder.WithAuthRefresh(options => { ... })` with `EnableAutoRefresh`, `RefreshBeforeExpiration`, and `OnRefreshed` / `OnRefreshFailed` callbacks.

**Where we'd love input**

* Are there scenarios here that wouldn't be covered by this approach (specific transports, auth schemes, Blazor Server, multi-server/backplane setups)?
* Is the client-driven TTL countdown the right model, or do you need server-initiated refresh requests?
* How important is the policy hook/permission-change handling versus just keeping the connection alive past token expiry?

Let us know how much that matters for your scenarios.

**Comment by thirstyape** [comment]
This sounds quite good. 

Would `options.AccessTokenProvider => MyCustomTokenRefresher();` fire when the current token is about to expire?

**Comment by BrennanConroy** [comment]
> Would `options.AccessTokenProvider => MyCustomTokenRefresher();` fire when the current token is about to expire?

Correct, right before calling the `/refresh` endpoint, the user provided access token factory is called to get a new auth token.

**Comment by gregoriusus** [comment]
Thank you for starting to resolve this issue.

If signalr is busy (a lot of messages), is there a possibility that some messages are lost during refresh process?

**Comment by Skyppid** [comment]
@BrennanConroy Wonderful, that looks pretty good already. I think our cases would be covered by that. The client-driven TTL is fine for most cases I'd say and adding server-initiated refresh requests should be quite easy to add later on in upcoming .NET versions if really required.

So far our workaround was also just a client-driven TTL which informs the server about a reconnect so it can flag and keep state-data of the user and then reconnects with a new token so the server can restore the state to the new connection. Pretty much the same design I'd say and it works decent enough but was quite tedious to build.

Not sure if this is already possible, but an option to close connections once tokens expired on server side would be nice. In case the client forgets the server would keep track of each connection's token expiry and close those that expired.

**Comment by BrennanConroy** [comment]
> If signalr is busy (a lot of messages), is there a possibility that some messages are lost during refresh process?

Nope, the current design keeps the connection open and working the whole time.

------------

> Not sure if this is already possible, but an option to close connections once tokens expired on server side would be nice.

That exists today:
https://learn.microsoft.com/aspnet/core/signalr/configuration?view=aspnetcore-10.0&tabs=dotnet#configure-advanced-http-options
> CloseOnAuthenticationExpiration

**Comment by BrennanConroy** [comment]
The initial implementation of this feature is now available in the nightly builds (main or preview6 column when that appears)
[dotnet/docs/builds-table.md at main · dotnet/dotnet](https://github.com/dotnet/dotnet/blob/main/docs/builds-table.md)

It will also be part of preview6 which is releasing July 14th.

Currently only implemented in .NET Client, TS Client coming in preview7 hopefully. Azure SignalR Service support is also in-progress.

Try it out and let us know how it works, what your scenario is, and whether it works for your scenario or not. Even if the feedback is `"this works great, I'm doing xyz"` that's a good signal.

General instructions for use below, need to configure server to allow authentication refresh per hub endpoint, and client is on by default but can also be configured.

.NET Client side:
```csharp
new HubConnectionBuilder().WithAuthenticationRefresh(new ()
{
    // optional
    EnableAutoRefresh = ...;,
    RefreshBeforeExpiration = TimeSpan.FromMinutes(1),
    OnAuthenticationRefreshed = ...;
    OnAuthenticationRefreshFailed = ...;
});
```

Server side:
```csharp
endpoints.MapHub<MyHub>("/hub", o =>
{
    o.EnableAuthenticationRefresh = true;

    // optional
    o.OnAuthenticationRefresh = ctx =>
    {
        return ValueTask.FromResult(true);
    };
});
```

**Comment by kujakipl** [comment]
@BrennanConroy 
Tested auth refresh in 11.0.100-preview.6 (.NET client)  works great.

Setup: WinForms/WPF clients + ASP.NET Core hub, custom auth pipeline, JWT TTL 60 s, RefreshBeforeExpiration = 15 s, WebSockets, EnableAuthenticationRefresh = true.
Results: 4 consecutive refresh cycles, each at exp 15 s; connection ID stable, no reconnects while original tokens expired; a server-side role change propagated to Context.User on the next refresh without dropping the connection.

Only note: claim changes propagate within one token TTL  worth mentioning in docs alongside the existing guidance. Looking forward to the TS client in preview 7.

**Comment by Youssef1313** [comment]
@BrennanConroy Is there any missing work here that needs to happen in 11 and should be tracked for 11.0-rc2?

**Comment by BrennanConroy** [comment]
The main feature work for this issue has been completed and is in 11.0, so I will be closing this issue. Any bugs, feature enhancements, or other issues should be raised as a new issue in this repository.

Docs for this feature at https://learn.microsoft.com/aspnet/core/signalr/authn-and-authz?view=aspnetcore-11.0#authentication-refresh, if you think there could be some documentation improvements file those at https://github.com/dotnet/aspnetcore.Docs.

Azure SignalR Service support is in-progress, you can follow its progress at https://github.com/Azure/azure-signalr/issues/2282

Thank you all for your feedback and patience.
